lib/web/ui/usercontext.go

```diff
@@ -1,43 +1,91 @@
 package ui
 
 import (
+	"github.com/gravitational/teleport/lib/defaults"
 	"github.com/gravitational/teleport/lib/services"
-
-	"github.com/gravitational/trace"
+	"github.com/gravitational/teleport/lib/utils"
 )
 
+type access struct {
+	Read   bool `json:"read"`
+	Edit   bool `json:"edit"`
+	Create bool `json:"create"`
+	Delete bool `json:"remove"`
+}
+
+type userACL struct {
+	// Sessions defines access to recorded sessions
+	Sessions access `json:"sessions"`
+	// AuthConnectors defines access to auth.connectors
+	AuthConnectors access `json:"authConnectors"`
+	// Roles defined access to roles
+	Roles access `json:"roles"`
+	// TrustedClusters defined access to trusted clusters
+	TrustedClusters access `json:"trustedClusters"`
+	// SSH defined access to servers
+	SSHLogins []string `json:"sshLogins"`
+}
+
 type userContext struct {
 	// Name is this user name
 	Name string `json:"userName"`
 	// Email is this user email
 	Email string `json:"userEmail"`
-	// ACL is this user access control list
-	ACL RoleAccess `json:"userAcl"`
+	// Logins is this user available logins
+	Logins []string `json:"userLogins"`
+	// ACL contains user access control list
+	ACL userACL `json:"userAcl"`
 }
 
-// NewUserContext returns userContext
-func NewUserContext(user services.User, allRoles []services.Role) (*userContext, error) {
-	userRoles := user.GetRoles()
-	roleNamesMap := map[string]bool{}
-	for _, name := range userRoles {
-		roleNamesMap[name] = true
+func getLogins(roleSet services.RoleSet) []string {
+	allLogins := []string{}
+	for _, role := range roleSet {
+		logins := role.GetLogins(services.Allow)
+		allLogins = append(allLogins, logins...)
 	}
 
-	accessSet := []RoleAccess{}
-	for _, item := range allRoles {
-		if roleNamesMap[item.GetName()] {
-			uiRole, err := NewRole(item)
-			if err != nil {
-				return nil, trace.Wrap(err)
-			}
+	return utils.Deduplicate(allLogins)
+}
 
-			accessSet = append(accessSet, uiRole.Access)
+func hasAccess(roleSet services.RoleSet, ctx *services.Context, kind string, verbs ...string) bool {
+	for _, verb := range verbs {
+		err := roleSet.CheckAccessToRule(ctx, defaults.Namespace, kind, verb)
+		if err != nil {
+			return false
 		}
 	}
 
-	userACL := MergeAccessSet(accessSet)
+	return true
+}
+
+func newAccess(roleSet services.RoleSet, ctx *services.Context, kind string) access {
+	return access{
+		Read:   hasAccess(roleSet, ctx, kind, services.VerbList),
```

Comment:
Looks like you are checking `VerbList` and granting access to `Read` here, was this intentional?

Reply:
Good catch! Will fix.